### ISSUE TITLE
Ray cannot identify non-base 10 GPU numbering on SLURM

### ISSUE THREAD

**Issue by yutaizhou** [ray-project/ray#8313]
Hey all,

I am using Ray on SLURM equipped with 2 GPUs per node. I get the following error when I submit a job that acquired GPUs:

```
                                                                                                                                                                   
  1   /state/partition1/user/yutai/raytmp
  2   Traceback (most recent call last):
  3   File "/home/gridsan/yutai/.conda/envs/football2/bin/ray", line 5, in <module>
  4     from ray.scripts.scripts import main
  5   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/__init__.py", line 65, in <module>
  6     from ray.worker import (
  7   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/worker.py", line 488, in <module>
  8     global_worker = Worker()
  9   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/worker.py", line 118, in __init__
 10     self.original_gpu_ids = ray.utils.get_cuda_visible_devices()
 11   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/utils.py", line 288, in get_cuda_visible_devices
 12     return [int(i) for i in gpu_ids_str.split(",")]
 13   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/utils.py", line 288, in <listcomp>
 14     return [int(i) for i in gpu_ids_str.split(",")]
 15 srun: error: d-10-14-2: task 0: Exited with exit code 1
 16 ValueError: invalid literal for int() with base 10: 'GPU-77e3a4cd-22fb-5d31-b8a6-409df6133eeb'
 17 adding workers
 18 GPUS: GPU-77e3a4cd-22fb-5d31-b8a6-409df6133eeb,GPU-3de6a983-6da7-e1cf-ffc6-a285c71ff5d3
 19 Traceback (most recent call last):
 20   File "ray_test.py", line 5, in <module>
 21     import ray
 22   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/__init__.py", line 65, in <module>
 23     from ray.worker import (
 24   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/worker.py", line 488, in <module>
 25     global_worker = Worker()
 26   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/worker.py", line 118, in __init__
 27     self.original_gpu_ids = ray.utils.get_cuda_visible_devices()
 28   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/utils.py", line 288, in get_cuda_visible_devices
 29     return [int(i) for i in gpu_ids_str.split(",")]
 30   File "/home/gridsan/yutai/.conda/envs/football2/lib/python3.6/site-packages/ray/utils.py", line 288, in <listcomp>
 31     return [int(i) for i in gpu_ids_str.split(",")]
 32 ValueError: invalid literal for int() with base 10: 'GPU-77e3a4cd-22fb-5d31-b8a6-409df6133eeb'
```

`echo "GPUS: $CUDA_VISIBLE_DEVICES"` gets me the following:
`GPUS: GPU-77e3a4cd-22fb-5d31-b8a6-409df6133eeb,GPU-3de6a983-6da7-e1cf-ffc6-a285c71ff5d3`

I am using Ray 0.8.4, and Python 3.6.10, on Ubuntu.

### Reproduction (REQUIRED)
Here are my sbatch shell script and python script. Note there are small deviations from this SLURM example in the docs as our SLURM system team initially got the example to work and I am simply using their modifications. The conda environment shouldn't be the bottleneck here. 
```
#!/bin/bash

#SBATCH -o %j.log
#SBATCH --job-name=ray_test
#SBATCH -n 4
#SBATCH -N 2
#SBATCH --gres=gpu:volta:2

export LC_ALL=C.UTF-8
export LANG=C.UTF-8

((worker_num=$SLURM_NNODES-1)) # Must be one less that the total number of nodes
echo $worker_num

# Set up environment
eval "$(conda shell.bash hook)"
conda activate football2
# or
# conda activate /home/gridsan/groups/ERGO_GRF/football_env/football

nodes=$(scontrol show hostnames $SLURM_JOB_NODELIST) # Getting the node names
nodes_array=( $nodes )

node1=${nodes_array[0]}

# Set IP address/port of head node
ip_prefix=$(srun --nodes=1 --ntasks=1 -w $node1 hostname --ip-address) # Making address
port='6379'
ip_head=$ip_prefix':'$port
export ip_head # Exporting for latter access by trainer.py

# Temporary directory for logging
tmpdir='/state/partition1/user/'$USER'/raytmp'
echo $tmpdir
mkdircmd='mkdir -p '$tmpdir

# Start the head
srun --nodes=1 --ntasks=1 -w $node1 $mkdircmd
srun --nodes=1 --ntasks=1 -w $node1 ray start --temp-dir=$tmpdir --block --head --redis-port=$port &
sleep 5

# Start workers
echo "adding workers"
for ((  i=1; i<=$worker_num; i++ ))
do
    node2=${nodes_array[$i]}
    srun --nodes=1 --ntasks=1 -w $node2 mkdir -p $tmpdir
    srun --nodes=1 --ntasks=1 -w $node2 ray start --temp-dir=$tmpdir --block --address=$ip_head & # Starting the workers
done
echo "GPUS: $CUDA_VISIBLE_DEVICES"
python ray_test.py $SLURM_NTASKS # Pass the total number of allocated CPUs
```

```
# trainer.py
import os
import sys
import time
import ray
import torch 


ray.init(address=os.environ["ip_head"])

@ray.remote
def f():
  time.sleep(1)

# The following takes one second (assuming that ray was able to access all of the allocated nodes).
start = time.time()
num_cpus = int(sys.argv[1])
ray.get([f.remote() for _ in range(num_cpus)])
end = time.time()
print(end - start)

```

**Comment by richardliaw** [comment]
Thanks for opening this issue @yutaizhou ! Is it possible to map the string back to the original device index (integers)?

**Comment by yutaizhou** [comment]
@richardliaw thanks for checking in! I think I could, since every node in our HPC system only has 2 GPUs anyway. I could just map the 2 GPUs as 0 and 1? Not sure if that would work across the entire system though. I will try it

**Comment by yutaizhou** [comment]
@richardliaw So I have tried remapping the GPU IDs but didn't get very far.. 
`srun --nodes=1 --ntasks=1 -w $node1 --export=ALL,CUDA_VISIBLE_DEVICES=0,1, ray start --temp-dir=$tmpdir --block --head --redis-port=$port &`

After talking to the HPC team, here is my understanding of the status and some direct quotes from them:

- "We’re using NVIDIA’s UUID API to set the device names to the UUID, rather than the default. The problem with the default naming scheme (0,1,etc) is that it is not consistent. What’s listed as GPU 0 might change even within a job, which you can imagine would cause major problems if you have two people on a node, each allocated one GPU." -MITSC

- In a shared environment like a HPC system, UUID seems to be the best (only?) way to to make sure people only have access to the GPUs that is allocated to them. Apparently TF and PyTorch don't have a problem with that. The HPC team essentially recommends Ray to accept non-default GPU naming scheme.

What do you think?

**Comment by richardliaw** [comment]
OK got it. Let me see if we can prio this. cc @anabranch

**Comment by yutaizhou** [comment]
Great, thank you Richard!

**Comment by Graeme22** [comment]
@yutaizhou Did you find a fix? I'm having the same issue on PBS.

**Comment by yutaizhou** [comment]
@Graeme22 yes but it is very hacky, and I still believe this should be addressed upstream in Ray.

When you do srun to spin up nodes, add an extra unset.sh file:

```
srun --nodes=1 --ntasks=1 -w $node1 unset.sh && ray start --temp-dir=$tmpdir --block --head --redis-port=$port &
```

unset.sh
```
unset CUDA_VISIBLE_DEVICES
```

**Comment by Graeme22** [comment]
For anyone having this issue, my quick fix was the following: Change line 288 of utils.py to this:

return gpu_ids_str.split(",")

The string based IDs still work as expected, and avoid casting error when they're not numerical.

**Comment by richardliaw** [comment]
This should be fixed in #9744! Please ping (or open a new issue if not!)
